content/blog/buildkite-cd.md

```diff
@@ -0,0 +1,136 @@
+
+We wanted a way to
+* Reduce the barrier to deployment making it easy and quick to encourage frequent integration and deployment
+* Deploy automatically to a staging server we can always review the state of the current master/main
+* Allow science to redeploy and see their code changes without having to go through the development team
+
+This blog post covers how we have used Buildkite within RESIDE to support continuous delivery of new features to staging environments. This does not cover details of the deployment script itself but how to set up Buildkite agents and configure a pipeline so that you can deploy via running a build, trigger it automatically and deploy specific tags of your docker images through environment variables.
+
+# Setting up Buildkite agent
+
+We want to create a separate agent from our normal build agents so that a deployment does not have to wait on any long-running builds. Buildkite agents can be configured to listen to specific queues. The default queue is called `default`, jobs added without a specified queue go onto the `default` queue and Buildkite agents without a specified queue pull jobs from this queue. We can create an agent which listens to a different queue by setting the `queue` tag in the [agent configuration](https://buildkite.com/docs/agent/v3/configuration). For example set `tags="queue=deploy"` to pull jobs from the `deploy` queue.
+
+To run the script on the remote server the agent will need an ssh key pair. These need to be static and not change if the agent is torn down and brought up again. We have added a persistent ssh key pair into our [vault](https://www.vaultproject.io/) secret store. Then when the agent is brought up we read the secret out of the vault and write it into the ssh dir.
+
+```
+AGENT_SSH=~buildkite-agent/.ssh
+mkdir -p $AGENT_SSH
+vault read -field=public secret/deploy/ssh > $AGENT_SSH/id_deploy.pub
+vault read -field=private secret/deploy/ssh > $AGENT_SSH/id_deploy
+chmod 600 $AGENT_SSH/id_deploy
+chown -R buildkite-agent.buildkite-agent $AGENT_SSH
+```
+
+We then add the public key to the list of authorised keys on the app host server.
+
+We can now start the agent. In the agents list on Buildkite, you should see the agent running with the queue set to `deploy`. We now have an agent listening on a dedicated deploy queue that can run scripts on the app host server. Next, we need to set up the deployment pipelines themselves.
+<img src="/img/buildkite-cd-agent.png" alt="png of agent list"/>
+
+# Deployment pipeline
+
+Create a repository for your deployment pipelines and add a pipeline yml at `./buildkite/deploy-pipeline.yml` with content
```

Comment:
Minor point, but the mode here seems to have changed from what "we" do, to instructions which "you" should follow. Should probably be consistent.